File: crashlytics/proj.android_studio/src/main/java/com/ee/plugin/CrashlyticsProtocol.java
package com.ee.plugin;

import android.content.Context;
import android.content.Intent;
import android.support.annotation.NonNull;
import android.util.Log;

import com.crashlytics.android.Crashlytics;
import com.crashlytics.android.core.BuildConfig;
import com.crashlytics.android.core.CrashTest;
import com.crashlytics.android.ndk.CrashlyticsNdk;
import com.ee.core.PluginProtocol;

import java.util.Map;

import io.fabric.sdk.android.Fabric;

/**
 * Created by Zinge on 7/4/16.
 */
public class CrashlyticsProtocol extends PluginProtocol {
    public CrashlyticsProtocol(Context context) {
        super(context);
        Fabric.Builder builder = new Fabric.Builder(context)
            .debuggable(BuildConfig.DEBUG)
            .logger(new CrashlyticsLogger(BuildConfig.DEBUG ? Log.DEBUG : Log.INFO))
            .kits(new Crashlytics(), new CrashlyticsNdk());

        Fabric.with(builder.build());
    }

    @NonNull
    @Override
    public String getPluginName() {
        return "CrashlyticsProtocol";
    }

    @Override
    public boolean onActivityResult(int requestCode, int responseCode, Intent data) {
        return false;
    }

    @Override
    public void onStart() {
    }

    @Override
    public void onStop() {
    }

    @Override
    public void onResume() {
    }

    @Override
    public void onPause() {
    }

    @Override
    public void onDestroy() {
    }

    @Override
    public boolean onBackPressed() {
        return false;
    }

    public void causeCrash() {
        Crashlytics.getInstance().crash();
    }

    public void causeException() {
        (new CrashTest()).throwRuntimeException("This is a test exception!");
    }

    public void log(@NonNull Map<String, Object> dict) {
        assert (dict.size() == 3);

        Integer priority = (Integer) dict.get("priority");
        String tag = (String) dict.get("tag");
        String message = (String) dict.get("message");

        assert (priority != null);
        assert (tag != null);
        assert (message != null);

        Crashlytics.log(priority, tag, message);
    }

    public void setString(@NonNull Map<String, Object> dict) {
        assert (dict.size() == 2);

        String key = (String) dict.get("key");
        String value = (String) dict.get("value");

        assert (key != null);
        assert (value != null);

        Crashlytics.setString(key, value);
    }

    public void setBool(@NonNull Map<String, Object> dict) {
        assert (dict.size() == 2);

        String key = (String) dict.get("key");
        Boolean value = (Boolean) dict.get("value");

        assert (key != null);
        assert (value != null);

        Crashlytics.setBool(key, value);
    }

    public void setInt(@NonNull Map<String, Object> dict) {
        assert (dict.size() == 2);

        String key = (String) dict.get("key");
        Integer value = (Integer) dict.get("value");

        assert (key != null);
        assert (value != null);

        Crashlytics.setInt(key, value);
    }

    public void setUserIdentifier(@NonNull Map<String, Object> dict) {
        assert (dict.size() == 1);

        String identifier = (String) dict.get("identifier");

        assert (identifier != null);

        Crashlytics.setUserIdentifier(identifier);
    }

    public void setUserName(@NonNull Map<String, Object> dict) {
        assert (dict.size() == 1);

        String name = (String) dict.get("name");

        assert (name != null);

        Crashlytics.setUserName(name);
    }

    public void setUserEmail(@NonNull Map<String, Object> dict) {
        assert (dict.size() == 1);

        String email = (String) dict.get("email");

        assert (email != null);

        Crashlytics.setUserName(email);
    }
}

File: crashlytics/src/ee/CrashlyticsProtocol.cpp
//
//  CrashlyticsProtocol.cpp
//  ee-crashlytics
//
//  Created by Zinge on 7/4/16.
//
//

#include "ee/CrashlyticsProtocol.hpp"

#include <ee/PluginProtocol.hpp>
#include <ee/LogLevel.hpp>
#include <nlohmann/json.hpp>

namespace ee {
namespace crashlytics {
CrashlyticsProtocol::CrashlyticsProtocol()
    : protocol_{new core::PluginProtocol{"CrashlyticsProtocol"}} {}

CrashlyticsProtocol::~CrashlyticsProtocol() = default;

void CrashlyticsProtocol::causeCrash() const {
    protocol_->callNative("causeCrash");
}

void CrashlyticsProtocol::causeException() const {
    protocol_->callNative("causeException");
}

void CrashlyticsProtocol::log(const core::LogLevel& level,
                              const std::string& tag,
                              const std::string& message) const {
    nlohmann::json json;
    json["priorityDescription"] = level.desc;
    json["priority"] = level.priority;
    json["tag"] = tag;
    json["message"] = message;
    protocol_->callNative("log", json.dump());
}

void CrashlyticsProtocol::setString(const std::string& key,
                                    const std::string& value) const {
    nlohmann::json json;
    json["key"] = key;
    json["value"] = value;
    protocol_->callNative("setString", json.dump());
}

void CrashlyticsProtocol::setBool(const std::string& key, bool value) const {
    nlohmann::json json;
    json["key"] = key;
    json["value"] = value;
    protocol_->callNative("setBool", json.dump());
}

void CrashlyticsProtocol::setInt(const std::string& key, int value) const {
    nlohmann::json json;
    json["key"] = key;
    json["value"] = value;
    protocol_->callNative("setInt", json.dump());
}

void CrashlyticsProtocol::setUserIdentifier(
    const std::string& identifier) const {
    nlohmann::json json;
    json["identifier"] = identifier;
    protocol_->callNative("setUserIdentifier", json.dump());
}

void CrashlyticsProtocol::setUserName(const std::string& name) const {
    nlohmann::json json;
    json["name"] = name;
    protocol_->callNative("setUserName", json.dump());
}

void CrashlyticsProtocol::setUserEmail(const std::string& email) const {
    nlohmann::json json;
    json["email"] = email;
    protocol_->callNative("setUserEmail", json.dump());
}
} // namespace crashlytics
} // namespace ee